tests/hvm_cpu.rs

```diff
@@ -0,0 +1,384 @@
+fn load_lang(file: &str) -> DefinitionBook {
+  let code = load_file(file);
+  parse_lang(&code)
+}
+
+// For every pair in the map, replaces all matches of a string the other string
+fn replace_template(mut code: String, map: &[(&str, &str)]) -> String {
+  for (from, to) in map {
+    code = code.replace(from, to);
+  }
+  code
+}
+
+trait Normal {
+  type Result;
+  fn normal(self, size: usize) -> Self::Result;
+}
+
+impl Normal for Book {
+  type Result = (run::Net, Net);
+
+  fn normal(self, size: usize) -> Self::Result {
+    let book = book_to_runtime(&self);
+    let mut rnet = run::Net::new(size);
+    rnet.boot(name_to_val("main"));
+    rnet.normal(&book);
+
+    let net = net_from_runtime(&rnet);
+    (rnet, net)
+  }
+}
+
+impl Normal for DefinitionBook {
+  type Result = (Term, DefNames, hvm_lang::RunInfo);
+
+  fn normal(self, size: usize) -> Self::Result {
+    hvm_lang::run_book(self, size).unwrap()
+  }
+}
+
+#[test]
+fn test_era_era() {
+  let net = parse_core("@main = * & * ~ *");
+  let (rnet, net) = net.normal(16);
+  assert_snapshot!(show_net(&net), @"*");
+  assert_debug_snapshot!(rnet.rewrites(), @"2");
+}
+
+#[test]
+fn test_con_dup() {
+  let net = parse_core("@main = root & (x x) ~ {2 * root}");
+  let (rnet, net) = net.normal(16);
+  assert_snapshot!(show_net(&net), @"(b b)");
+  assert_debug_snapshot!(rnet.rewrites(), @"5");
+}
+
+#[test]
+fn test_bool_and() {
+  let book = parse_core(
+    "
+    @true = (b (c b))
+    @fals = (b (c c))
+    @and  = ((b (@fals c)) (b c))
+    @main = root & @and ~ (@true (@fals root))
+  ",
+  );
+  let (rnet, net) = book.normal(64);
+
+  assert_snapshot!(show_net(&net), @"(b (c c))");
+  assert_debug_snapshot!(rnet.rewrites(), @"8");
+}
+
+#[test]
+fn test_church_mul() {
+  let book = load_lang("church_mul.hvm");
+  let (term, defs, info) = book.normal(64);
+
+  assert_snapshot!(show_net(&info.net), @"({2 ({2 b c} d) {3 (d e) (e {2 c f})}} (b f))");
+  assert_snapshot!(term.to_string(&defs), @"λa λb (a (a (a (a (a (a b))))))");
+}
+
+#[test]
+fn test_neg_fusion() {
+  let book = load_lang("neg_fusion.hvm");
+  let (_, _, info) = book.normal(512);
+
+  assert_snapshot!(show_net(&info.net), @"(b (* b))");
+  assert_debug_snapshot!(info.stats.rewrites.total_rewrites(), @"153");
+}
+
+#[test]
+fn test_tree_alloc() {
+  let book = load_lang("tree_alloc.hvm");
+  let (_, _, info) = book.normal(512);
+
+  assert_snapshot!(show_net(&info.net), @"(b (* b))");
+  assert_debug_snapshot!(info.stats.rewrites.total_rewrites(), @"104");
+}
+
+fn make_queue(len: u32) -> DefinitionBook {
+  let template = load_file("queue.hvm");
+  let mut body = String::new();
+
+  for i in 1 ..= len {
+    body += &format!("let q = (Qadd {i} q)\n")
+  }
+
+  for i in 0 .. len {
+    body += &format!("(Qrem q λv{i} λq\n");
+  }
+
+  for i in 1 ..= len {
+    body += &format!("(Cons {i} ");
+  }
+
+  body += &format!(" Nil{}", ")".repeat(len as usize * 2));
```

Comment:
```suggestion
  let mut body = String::new();
  fn add_to_queue(mut elements: Vec<u32>) -> String {
    match elements.pop() {
      None => "".to_string(),
      Some(head) => format!("let q = (Qadd {} q)\n{}", head, &add_to_queue(elements)),
    }
  }
  fn rm_from_queue(n: u32, body: String) -> String {
    match n {
      0 => body,
      n => format!("(Qrem q λv{} λq {})", n, &rm_from_queue(n - 1, body)),
    }
  }
  fn to_list(mut elements: Vec<String>) -> String {
    match elements.pop() {
      None => "Nil".to_string(),
      Some(head) => format!("(Cons {} {})", head, &to_list(elements)),
    }
  }

  // any elements in the queue
  let queue_els = (1..=len).into_iter().map(|n| n * 5).collect();
  body += &add_to_queue(queue_els);
  
  // creating a list from all of the removed variables in the queue
  let var_list = (1..=len).into_iter().map(|n| format!("v{n}")).collect();
  body += &rm_from_queue(len, to_list(var_list));
```

Reply:
notice that the original test https://github.com/HigherOrderCO/HVM/blob/master/examples/queue/main.hvm#L21 uses the variables to construct the list

Reply:
also, maybe rename `Qrem` to `Qrm`/`Q.rm`/`Queue.rm` or `Qremove`/`Q.remove`/`Queue.remove`.

I first thought `Qrem` was queue remainder, and I was wondering if it was an algorithm I didn't know

Reply:
by the way, the more I think about it, do we really need to create an arbitrarily-sized queue? can't we just have a fixed one in the tests? it'll make the tests easier to read, and I don't think it will hinder the correctness checking

Reply:
> notice that the original test https://github.com/HigherOrderCO/HVM/blob/master/examples/queue/main.hvm#L21 uses the variables to construct the list

I got this version from hvm_lang [tests](https://github.com/HigherOrderCO/hvm-lang/blob/master/tests/golden_tests/run_single_files/queue.hvm)

Reply:
> by the way, the more I think about it, do we really need to create an arbitrarily-sized queue? can't we just have a fixed one in the tests? it'll make the tests easier to read, and I don't think it will hinder the correctness checking

Yeah, I agree, I'll make it static then